contributors/design-proposals/network/service-topology-key.md

```diff
@@ -0,0 +1,171 @@
+
+Users can control the level of topology. For example, if someone run logging agent as a daemonset, he can set the "hard" topology requirement for same-host. If "hard" is not met, then just return "service not available". 
+
+And if someone set a "soft" topology requirement for same-host, say he "preferred" same-host endpoints and can accept other hosts when for some reasons local service's backend is not available on some host.
+
+If multiple endpoints satisfy the "hard" or "soft" topology requirement, we will randomly pick one by default. Routing decision is expected to be implemented in L3/4 VIP level such as kube proxy.
+
+## Implementation details
+
+### API changes
+
+#### Service API changes
+
+As with scheduling topology, there are currently two types of topology, called `Required` and `Preferred` which denote “hard” vs. “soft” requirements. 
+
+Similarly, we introduce a topology mode and an associated topology key in service topology.
+
+```go
+type Service struct {
+  Spec:
+    // topology is used to achieve "local" service in a given topology level.
+    // User can control what ever topology level they want.
+    // +optional
+    Topology ServiceTopology `json:"topology" protobuf:"bytes,1,opt,name=topology"`
+}
+
+// Defines a service topolgoy information.
+type ServiceTopology struct {
+    // Valid values for mode are "ignored", "required", "preferred".
+    // "ignored" is the default value and the associated topology key will have no effect.
+    // "required" is the "hard" requirement for topology key and an example would be  “only visit service backends in the same zone”. 
+    // If the topology requirements specified by this field are not met, the LB, such as kube-proxy will not pick endpoints for the service.
+    // "preferred" is the "soft" requirement for topology key and an example would be   
+    // "prefer to visit service backends in the same rack, but OK to other racks if none match"
+    // +optional
+    Mode ServicetopologyMode `json:"mode" protobuf:"bytes,1,opt,name=mode"`
+    // key is the key for the node label that the system uses to denote    
+    // such a topology domain. There are some built-in topology keys, e.g. 
+    // kubernetes.io/hostname, failure-domain.beta.kubernetes.io/zone and failure-domain.beta.kubernetes.io/region etc.
+    // The built-in topology keys can be good examples and we recommend users switch to a similar mode for portability, but it's NOT enforced.
+    // Users can define whatever topolgoy key they like since topology is arbitrary.
+    // +optional
+    Key string `json:"key" protobuf:"bytes,2,opt,name=key"`
+}
+
+type ServicetopologyMode string
+```
+
+An example of service that uses topologyKey:
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: with-service-topology
+spec:
+  topology:
+    key: kubernetes.io/hostname
+    mode: required
```

Comment:
Constants are always `CamelCase`

Reply:
There is an impl. on this proposal, please check: https://github.com/kubernetes/kubernetes/pull/58732